File: backend/services/GroceryListService.ts
/**
 * GroceryListService — Fase 5.2 (PLAN_NUTRICION)
 *
 * Consolida ítems de un plan diario en una lista de mercado inteligente
 * para N días, agrupando por food_id y costando con precios TCA 2018.
 */

export interface GroceryPlanItem {
  food_id: string;
  name?: string;
  /** Porción en gramos de UNA jornada (suma de todas las comidas del día). */
  portion_g: number;
  purchase_unit?: number | string | null;
  purchase_price?: number | null;
  /** Categoría opcional para ordenar (ej. proteína, cereal). */
  category?: string | null;
}

export interface GroceryListLine {
  food_id: string;
  name: string;
  category: string | null;
  /** Gramos por día (plan base). */
  daily_portion_g: number;
  /** Gramos totales para el periodo (daily × days). */
  total_grams: number;
  /** Unidad de compra parseada (g). null si no hay dato TCA. */
  purchase_unit_g: number | null;
  /** Precio unitario TCA. */
  purchase_price: number | null;
  /**
   * Unidades de compra sugeridas (ceil(total_grams / purchase_unit_g)).
   * null si no se puede calcular.
   */
  units_to_buy: number | null;
  /** Costo = units_to_buy * purchase_price (o prorrateo por gramos si se prefiere). */
  line_cost: number | null;
  /** Gramos equivalentes a units_to_buy × unit (puede sobrar vs total_grams). */
  purchase_grams_rounded: number | null;
}

export interface SmartGroceryList {
  days: number;
  item_count: number;
  lines_with_price: number;
  lines_missing_price: number;
  /** Gran total de la canasta familiar ($). */
  grand_total: number;
  currency_note: string;
  lines: GroceryListLine[];
  generated_at: string;
}

function round(n: number, decimals = 2): number {
  const f = 10 ** decimals;
  return Math.round(n * f) / f;
}

/** Extrae gramos desde "500", "500 g", "500g o cc". */
export function parsePurchaseUnitGrams(raw: number | string | null | undefined): number | null {
  if (raw == null || raw === '') return null;
  if (typeof raw === 'number') {
    return Number.isFinite(raw) && raw > 0 ? raw : null;
  }
  const match = String(raw)
    .trim()
    .replace(',', '.')
    .match(/-?\d+(\.\d+)?/);
  if (!match) return null;
  const n = Number(match[0]);
  return Number.isFinite(n) && n > 0 ? n : null;
}

function asPrice(raw: number | null | undefined): number | null {
  if (raw == null) return null;
  const n = Number(raw);
  return Number.isFinite(n) && n >= 0 ? n : null;
}

/**
 * Genera la lista de mercado inteligente.
 *
 * @param planItems Ítems del plan de UN día (ya consolidados por comida o crudos).
 * @param days Número de días a comprar (default 7).
 *
 * Reglas de costo:
 * - total_grams = sum(portion_g por food_id) * days
 * - units_to_buy = ceil(total_grams / purchase_unit_g)
 * - line_cost = units_to_buy * purchase_price
 *   (compra por empaques enteros; refleja ticket real de supermercado)
 */
export function generateSmartGroceryList(
  planItems: GroceryPlanItem[],
  days = 7
): SmartGroceryList {
  if (!Number.isFinite(days) || days < 1) {
    throw new Error('days debe ser un entero ≥ 1.');
  }
  const periodDays = Math.floor(days);

  const byFood = new Map<
    string,
    {
      name: string;
      category: string | null;
      daily_portion_g: number;
      purchase_unit: number | string | null | undefined;
      purchase_price: number | null | undefined;
    }
  >();

  for (const item of planItems) {
    if (!item?.food_id) continue;
    const portion = Number(item.portion_g);
    if (!Number.isFinite(portion) || portion <= 0) continue;

    const prev = byFood.get(item.food_id);
    if (!prev) {
      byFood.set(item.food_id, {
        name: (item.name || item.food_id).trim(),
        category: item.category?.trim() || null,
        daily_portion_g: portion,
        purchase_unit: item.purchase_unit,
        purchase_price: item.purchase_price,
      });
    } else {
      prev.daily_portion_g += portion;
      if (!prev.name && item.name) prev.name = item.name.trim();
      if (!prev.category && item.category) prev.category = item.category.trim();
      // Preferir el primer precio/unidad no nulo
      if (prev.purchase_unit == null && item.purchase_unit != null) {
        prev.purchase_unit = item.purchase_unit;
      }
      if (prev.purchase_price == null && item.purchase_price != null) {
        prev.purchase_price = item.purchase_price;
      }
    }
  }

  const lines: GroceryListLine[] = [];
  let grandTotal = 0;
  let withPrice = 0;
  let missingPrice = 0;

  for (const [food_id, agg] of byFood.entries()) {
    const totalGrams = round(agg.daily_portion_g * periodDays, 1);
    const unitG = parsePurchaseUnitGrams(agg.purchase_unit);
    const price = asPrice(agg.purchase_price ?? null);

    let unitsToBuy: number | null = null;
    let lineCost: number | null = null;
    let purchaseGramsRounded: number | null = null;

    if (unitG != null && price != null) {
      unitsToBuy = Math.max(1, Math.ceil(totalGrams / unitG));
      purchaseGramsRounded = round(unitsToBuy * unitG, 1);
      lineCost = round(unitsToBuy * price, 2);
      grandTotal += lineCost;
      withPrice += 1;
    } else if (unitG != null && price == null) {
      unitsToBuy = Math.max(1, Math.ceil(totalGrams / unitG));
      purchaseGramsRounded = round(unitsToBuy * unitG, 1);
      missingPrice += 1;
    } else if (price != null && unitG == null) {
      // Sin unidad: prorrateo asumiendo precio por 100 g (fallback clínico)
      lineCost = round((totalGrams / 100) * price, 2);
      grandTotal += lineCost;
      withPrice += 1;
    } else {
      missingPrice += 1;
    }

    lines.push({
      food_id,
      name: agg.name,
      category: agg.category,
      daily_portion_g: round(agg.daily_portion_g, 1),
      total_grams: totalGrams,
      purchase_unit_g: unitG,
      purchase_price: price,
      units_to_buy: unitsToBuy,
      line_cost: lineCost,
      purchase_grams_rounded: purchaseGramsRounded,
    });
  }

  lines.sort((a, b) => {
    const ca = (a.category || 'zzz').localeCompare(b.category || 'zzz', 'es');
    if (ca !== 0) return ca;
    return a.name.localeCompare(b.name, 'es');
  });

  return {
    days: periodDays,
    item_count: lines.length,
    lines_with_price: withPrice,
    lines_missing_price: missingPrice,
    grand_total: round(grandTotal, 2),
    currency_note: 'COP / moneda de purchase_price en TCA 2018',
    lines,
    generated_at: new Date().toISOString(),
  };
}

export const GroceryListService = {
  parsePurchaseUnitGrams,
  generateSmartGroceryList,
} as const;

export default GroceryListService;

File: backend/internal/core/services/grocery_list_service.go
package services

import (
	"fmt"
	"math"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/kinesys/clinical-platform-backend/internal/core/domain"
	"github.com/kinesys/clinical-platform-backend/internal/core/ports"
)

type groceryListService struct{}

func NewGroceryListService() ports.GroceryListService {
	return &groceryListService{}
}

func (s *groceryListService) ParsePurchaseUnitGrams(raw *string) *float64 {
	if raw == nil || *raw == "" {
		return nil
	}

	str := strings.TrimSpace(*raw)
	str = strings.ReplaceAll(str, ",", ".")

	re := regexp.MustCompile(`-?\d+(\.\d+)?`)
	match := re.FindString(str)
	if match == "" {
		return nil
	}

	val, err := strconv.ParseFloat(match, 64)
	if err != nil || math.IsNaN(val) || math.IsInf(val, 0) || val <= 0 {
		return nil
	}

	return &val
}

func parsePrice(raw *float64) *float64 {
	if raw == nil {
		return nil
	}
	if math.IsNaN(*raw) || math.IsInf(*raw, 0) || *raw < 0 {
		return nil
	}
	val := *raw
	return &val
}

type foodAggregation struct {
	Name          string
	Category      *string
	DailyPortionG float64
	PurchaseUnit  *string
	PurchasePrice *float64
}

func (s *groceryListService) GenerateSmartGroceryList(planItems []domain.GroceryPlanItem, days int) (domain.SmartGroceryList, error) {
	if days < 1 {
		return domain.SmartGroceryList{}, fmt.Errorf("days debe ser un entero >= 1")
	}

	byFood := make(map[string]*foodAggregation)

	for _, item := range planItems {
		if item.FoodID == "" {
			continue
		}
		portion := item.PortionG
		if math.IsNaN(portion) || math.IsInf(portion, 0) || portion <= 0 {
			continue
		}

		agg, exists := byFood[item.FoodID]
		if !exists {
			name := item.FoodID
			if item.Name != nil && strings.TrimSpace(*item.Name) != "" {
				name = strings.TrimSpace(*item.Name)
			}
			var category *string
			if item.Category != nil && strings.TrimSpace(*item.Category) != "" {
				c := strings.TrimSpace(*item.Category)
				category = &c
			}

			byFood[item.FoodID] = &foodAggregation{
				Name:          name,
				Category:      category,
				DailyPortionG: portion,
				PurchaseUnit:  item.PurchaseUnit,
				PurchasePrice: item.PurchasePrice,
			}
		} else {
			agg.DailyPortionG += portion
			if agg.Name == item.FoodID && item.Name != nil && strings.TrimSpace(*item.Name) != "" {
				agg.Name = strings.TrimSpace(*item.Name)
			}
			if agg.Category == nil && item.Category != nil && strings.TrimSpace(*item.Category) != "" {
				c := strings.TrimSpace(*item.Category)
				agg.Category = &c
			}
			if agg.PurchaseUnit == nil && item.PurchaseUnit != nil {
				agg.PurchaseUnit = item.PurchaseUnit
			}
			if agg.PurchasePrice == nil && item.PurchasePrice != nil {
				agg.PurchasePrice = item.PurchasePrice
			}
		}
	}

	var lines []domain.GroceryListLine
	var grandTotal float64
	var withPrice, missingPrice int

	for foodID, agg := range byFood {
		totalGrams := roundTo(agg.DailyPortionG*float64(days), 1)
		unitG := s.ParsePurchaseUnitGrams(agg.PurchaseUnit)
		price := parsePrice(agg.PurchasePrice)

		var unitsToBuy *int
		var lineCost *float64
		var purchaseGramsRounded *float64

		if unitG != nil && price != nil {
			u := int(math.Max(1, math.Ceil(totalGrams/(*unitG))))
			unitsToBuy = &u
			
			pGrams := roundTo(float64(u)*(*unitG), 1)
			purchaseGramsRounded = &pGrams
			
			lCost := roundTo(float64(u)*(*price), 2)
			lineCost = &lCost
			
			grandTotal += lCost
			withPrice++
		} else if unitG != nil && price == nil {
			u := int(math.Max(1, math.Ceil(totalGrams/(*unitG))))
			unitsToBuy = &u
			
			pGrams := roundTo(float64(u)*(*unitG), 1)
			purchaseGramsRounded = &pGrams
			
			missingPrice++
		} else if price != nil && unitG == nil {
			// Fallback clinico: prorrateo asumiendo precio por 100g
			lCost := roundTo((totalGrams/100.0)*(*price), 2)
			lineCost = &lCost
			grandTotal += lCost
			withPrice++
		} else {
			missingPrice++
		}

		lines = append(lines, domain.GroceryListLine{
			FoodID:               foodID,
			Name:                 agg.Name,
			Category:             agg.Category,
			DailyPortionG:        roundTo(agg.DailyPortionG, 1),
			TotalGrams:           totalGrams,
			PurchaseUnitG:        unitG,
			PurchasePrice:        price,
			UnitsToBuy:           unitsToBuy,
			LineCost:             lineCost,
			PurchaseGramsRounded: purchaseGramsRounded,
		})
	}

	sort.Slice(lines, func(i, j int) bool {
		catI := "zzz"
		if lines[i].Category != nil {
			catI = *lines[i].Category
		}
		catJ := "zzz"
		if lines[j].Category != nil {
			catJ = *lines[j].Category
		}

		if catI != catJ {
			return catI < catJ
		}
		return lines[i].Name < lines[j].Name
	})

	return domain.SmartGroceryList{
		Days:              days,
		ItemCount:         len(lines),
		LinesWithPrice:    withPrice,
		LinesMissingPrice: missingPrice,
		GrandTotal:        roundTo(grandTotal, 2),
		CurrencyNote:      "COP / moneda de purchase_price en TCA 2018",
		Lines:             lines,
		GeneratedAt:       time.Now().UTC().Format(time.RFC3339),
	}, nil
}
